### ISSUE TITLE
SQLFluff incorrectly flags valid CTE alias

### ISSUE THREAD

**Issue by ssd4tvg** [sqlfluff/sqlfluff#7502]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff incorrectly flags valid CTE alias

### Expected Behaviour

SQLFluff it should not raise ST09 error, when everything is correct

### Observed Behaviour

(base) sagemaker-user@default:~/user-default-efs/ds-retail-kvi-segmentation$ sqlfluff lint
== [/mnt/custom-file-systems/efs/fs-0756a467c33f1b541_fsap-090d54836e3617351/ds-retail-kvi-segmentation/pipeline/model/sql/sql_kvi_line.sql] FAIL                                                                                                           
L: 383 | P:  12 | ST09 | Joins should list the table referenced earlier first.
                       | [structure.join_condition_order]
All Finished 📜 🎉!   

### How to reproduce

fix this query -
select
        ld.line_number,
        count(distinct o.purchaseordernumber) as ordering_visits
    from prod_clickstream_views.orders_c as o
    inner join prod_clickstream_views.productpage_c as pp on o.sessionkey = pp.sessionkey
    inner join trading_decision_support.zz_sh_line_data_nh as ld
        on pp.productid = ld.line_number
    where
        cast(o.eventtimestamp as date) between (td_sunday(date)) - 27 and td_sunday(date)
        and pp.principalbrandweb = 'LEX'
        and ld.division in ('TOYS', 'BEAUTY', 'SPORTS AND LEISURE', 'ELECTRICAL')
    group by ld.line_number

### Dialect

NA

### Version

(base) sagemaker-user@default:~/user-default-efs/ds-retail-kvi-segmentation$ sqlfluff version
4.0.4

### Configuration

NA

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
Hi @ssd4tvg  - what kind of dialect is this SQL statement? Postgres? Oracle? Pls specify

**Comment by ssd4tvg** [comment]
Teradata

**Comment by peterbud** [comment]
I have tested the above query, and it does not trigger the ST09 rule with the latest version of `sqlfluff`. Also I'm not sure where is the CTE in this query.

```sql
select
ld.line_number,
count(distinct o.purchaseordernumber) as ordering_visits
from prod_clickstream_views.orders_c as o
inner join prod_clickstream_views.productpage_c as pp on o.sessionkey = pp.sessionkey
inner join trading_decision_support.zz_sh_line_data_nh as ld
on pp.productid = ld.line_number
where
cast(o.eventtimestamp as date) between (td_sunday(date)) - 27 and td_sunday(date)
and pp.principalbrandweb = 'LEX'
and ld.division in ('TOYS', 'BEAUTY', 'SPORTS AND LEISURE', 'ELECTRICAL')
group by ld.line_number
```

```bash
sqlfluff fix test.sql  -d teradata
==== finding fixable violations ====
== [test.sql] FAIL                                                                                                                  
L:   2 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:   3 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:   5 | P:   1 | LT05 | Line is too long (85 > 80). [layout.long_lines]
L:   7 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:   9 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:   9 | P:   1 | LT05 | Line is too long (85 > 80). [layout.long_lines]
L:  10 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:  11 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:  12 | P:  24 | LT12 | Files must end with a single trailing newline.
                       | [layout.end_of_file]
== [test.sql] FIXED
9 fixable linting violations found
```

and here is the fixed sql:
```sql
select
    ld.line_number,
    count(distinct o.purchaseordernumber) as ordering_visits
from prod_clickstream_views.orders_c as o
inner join
    prod_clickstream_views.productpage_c as pp
    on o.sessionkey = pp.sessionkey
inner join trading_decision_support.zz_sh_line_data_nh as ld
    on pp.productid = ld.line_number
where
    cast(o.eventtimestamp as date) between (td_sunday(date))
    - 27 and td_sunday(date)
    and pp.principalbrandweb = 'LEX'
    and ld.division in ('TOYS', 'BEAUTY', 'SPORTS AND LEISURE', 'ELECTRICAL')
group by ld.line_number
```
